qutip/tests/solver/heom/test_bofin_baths.py

```diff
@@ -463,3 +472,247 @@ def test_create(self):
+        assert np.isclose(pow, pow2).all()
+
+    def test_get_fit(self):
+        a, b, c = [list(range(5))] * 3
+        w = np.linspace(0.1, 10 * self.wc, 1000)
+        J = self.bath._spectral_density_approx(w, a, b, c)
+        self.bath.get_fit(J, w)
+        a2, b2, c2 = self.bath.params_spec
+        J2 = self.bath._spectral_density_approx(w, a2, b2, c2)
+        assert np.sum(J - J2) < (1e-12)
+        assert self.bath.spec_n <= 5
+
+    def test_matsubara_coefficients_from_spectral_fit(self):
+        ckAR, vkAr, ckAI, vkAI = (
+            [
+                (0.29298628613108785 - 0.2097848947562078j),
+                (0.29298628613108785 + 0.2097848947562078j),
+                (-0.01608448645347298 + 0j),
+                (-0.002015397620259254 + 0j),
+            ],
+            [(1 - 1j), (1 + 1j), (6.283185307179586 + 0j),
+             (12.566370614359172 + 0j)],
+            [(-0 + 0.25j), -0.25j],
+            [(1 - 1j), (1 + 1j)],
+        )
+        self.bath.params_spec = np.array([1]), np.array([1]), np.array([1])
+        self.bath._matsubara_spectral_fit()
+        ckAR2, vkAr2, ckAI2, vkAI2 = self.bath.cvars
+        assert np.isclose(ckAR, ckAR2).all()
+        assert np.isclose(vkAr, vkAr2).all()
+        assert np.isclose(ckAI, ckAI2).all()
+        assert np.isclose(vkAI, vkAI2).all()
+
+
+class TestFitCorr:
+    alpha = 0.005
+    wc = 1
+    T = 1
+    bath = FitCorr(sigmax())
+
+    def test_corr_approx(self):
+        t = np.array([1 / 10, 1 / 5, 1])
+        C = np.array(
+            [
+                -0.900317 + 0.09033301j,
+                -0.80241065 + 0.16265669j,
+                -0.19876611 + 0.30955988j,
+            ]
+        )
+        a, b, c = -np.array([list(range(2))] * 3)
+        C2 = self.bath._corr_approx(t, a, b, c)
+        assert np.isclose(C, C2).all()
+
+    def test_fit_correlation(self):
+        a, b, c = -np.array([list(range(2))] * 3)
+        t = np.linspace(0.1, 10 / self.wc, 1000)
+        C = self.bath._corr_approx(t, a, b, c)
+        self.bath.fit_correlation(t, C, Nr=3, Ni=3)
+        a2, b2, c2 = self.bath.params_real
+        ai2, bi2, ci2 = self.bath.params_imag
+        C2re = self.bath._corr_approx(t, a2, b2, c2)
+        C2imag = self.bath._corr_approx(t, ai2, bi2, ci2)
+        assert np.isclose(np.real(C), np.real(C2re)).all()
+        assert np.isclose(np.imag(C), np.imag(C2imag)).all()
+
+    def test_matsubara_coefficients(self):
+        ans = np.array(
+            [
+                [(0.5 + 0j), (0.5 + 0j), (0.5 - 0j), (0.5 - 0j)],
+                [(-1 - 1j), (-1 - 1j), (-1 + 1j), (-1 + 1j)],
+                [-0.5j, -0.5j, 0.5j, 0.5j],
+                [(-1 - 1j), (-1 - 1j), (-1 + 1j), (-1 + 1j)],
+            ]
+        )
+        self.bath.params_real = [[1] * 2] * 3
+        self.bath.params_imag = [[1] * 2] * 3
+        self.bath._matsubara_coefficients()
+        assert np.isclose(np.array(self.bath.cvars), ans).all()
```

Comment:
1. Don't the params to attributes like this. If the FitCorr needs parameters, pass them when initializing the object.
2. It's not great if tests call private `_` methods like `_matsubara_coefficients()`, although occasionally it's sensible. This does not look like such a case.
3. Definitely don't have methods make attributes like `cvars` appear. If the method needs to return a result, just return it.

Probably there is a deeper issue here, so the answer might not be to address directly the comments above, but just to change something so that all of the issues go away. I see @pmenczel noted that the `.cvars` were unused and should just be removed anyway?

Reply:
1) I understand, I did it like that because I didn't want to run a fit to test that function but rather wanted an identifiable fixed input, will rethink how to do it, or will simply fit
2)I wasn't aware, thanks !
3) understood, if the same thing is needed in many places then I just run the method in each other method that needs it, right?

Yes I guess the deeper issue was that I didn't really understand how the exponent construction and the combine works so I was using those cvars instead of the exponents, untill Paul suggested using exponents and removing cvars, however I didn't do a good job and they remained in a few places, sorry about that!